### PR TITLE
Adding the correlation log configuration support

### PR DESCRIPTION
**Pull request by wasuradananjith**
## Purpose
Fixes: wso2/api-manager#275
Fixes: wso2/api-manager#566

## Goals
Adding the correlation log support of APIM to APICTL

## Approach
Refactoring the draft PR in [1].
- Introduced a command to get correlation logging configurations of different components in an environment.
- Introduced a command to set correlation logging configurations of different components in an environment.

[1] https://github.com/wso2/product-apim-tooling/pull/930

### CODE REVIEW

File: import-export-cli/cmd/getCorrelationLogging.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	},
+}
+
+func executeGetCorrelationLoggingCmd(credential credentials.Credential) {
+	components, err := impl.GetCorrelationLogComponentListFromEnv(credential, getCorrelationLoggingEnvironment)
+
+	if err == nil {
+		impl.PrintCorrelationLoggers(components, getCorrelationLoggingCmdFormat)
+	} else {
+		utils.Logln(utils.LogPrefixError+"Getting list of correlation log configurations", err)
```
Comment (ashera96) [inline import-export-cli/cmd/getCorrelationLogging.go:59]: Shall we fix the formatting? 
```suggestion
		utils.Logln(utils.LogPrefixError + "Getting list of correlation log configurations", err)
```

File: import-export-cli/cmd/setCorrelationLogging.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+			utils.HandleErrorAndExit("Error getting credentials", err)
+		}
+		executeSetCorrelationLoggingCmd(cred)
+	},
+}
+
+func executeSetCorrelationLoggingCmd(credential credentials.Credential) {
+	resp, err := impl.SetCorrelationLoggingComponent(credential, setCorrelationLoggingEnvironment, setCorrelationLoggingComponentName, setCorrelationLoggingEnabled, setCorrelationLoggingDeniedThreads)
+	if err != nil {
+		utils.Logln(utils.LogPrefixError+"Setting the correlation log component configs", err)
```
Comment (ashera96) [inline import-export-cli/cmd/setCorrelationLogging.go:63]:
```suggestion
		utils.Logln(utils.LogPrefixError + "Setting the correlation log component configs", err)
```

File: import-export-cli/impl/logger.go [hunk 3]
```diff
@@ -204,3 +233,128 @@ func SetAPILoggingLevel(credential credentials.Credential, environment, apiId, t
...
+// GET Correlation Components List 
+func GetCorrelationLogComponentListFromEnv(credential credentials.Credential, environment string) (components []utils.CorrelationComponent, err error) {
+	// Base64 encoding the credentials
+	b64encodedCredentials := credentials.GetBasicAuth(credential)
+	// Prepping the headers
+	headers := make(map[string]string)
+	headers[utils.HeaderAuthorization] = utils.HeaderValueAuthBasicPrefix + " " + b64encodedCredentials
+	
+	correlationDevOpsEP := utils.GetCorrelationLoggingEndPointOfEnv(environment, utils.MainConfigFilePath)
+	utils.Logln(utils.LogPrefixInfo+"URL : " + correlationDevOpsEP)
```
Comment (ashera96) [inline import-export-cli/impl/logger.go:246]:
```suggestion
	utils.Logln(utils.LogPrefixInfo + "URL : " + correlationDevOpsEP)
```

File: import-export-cli/impl/logger.go [hunk 4]
```diff
@@ -204,3 +233,128 @@ func SetAPILoggingLevel(credential credentials.Credential, environment, apiId, t
...
+					cc.Properties[len(cc.Properties)-1].Value = strings.Split(deniedThreads, ",")
+				}
+			}
+		}
+		requestComponents = append(requestComponents, cc)
+	}
+	
+	b,err := json.Marshal(requestComponents)
+	if err != nil {
+		utils.Logln("Error when creating a json ")
```
Comment (ashera96) [inline import-export-cli/impl/logger.go:342]:
```suggestion
		utils.Logln("Error when creating a json")		
```

File: import-export-cli/impl/logger.go [hunk 5]
```diff
@@ -204,3 +233,128 @@ func SetAPILoggingLevel(credential credentials.Credential, environment, apiId, t
...
+	headers[utils.HeaderContentType] = utils.HeaderValueApplicationJSON
+	body := string(b)
+	body = "{\"components\":" + body + "}"
+	putResp, err := utils.InvokePutRequest(nil, correlationDevOpsEP, headers, body)
+
+	if err != nil {
+		utils.HandleErrorAndExit("Unable to connect to "+correlationDevOpsEP, err)
+	}
+
+	utils.Logln(utils.LogPrefixInfo+"PUT Response:", putResp.Status())
```
Comment (ashera96) [inline import-export-cli/impl/logger.go:355]:
```suggestion
	utils.Logln(utils.LogPrefixInfo + "PUT Response:", putResp.Status())
```

File: import-export-cli/impl/logger.go [hunk 6]
```diff
@@ -204,3 +233,128 @@ func SetAPILoggingLevel(credential credentials.Credential, environment, apiId, t
...
+	if putResp.StatusCode() == http.StatusOK {
+		return putResp, nil
+	}
+	return nil, errors.New(string(putResp.Body()))
+}
```
Comment (ashera96) [inline import-export-cli/impl/logger.go:360]: Shall we add a newline at the EOF?

File: import-export-cli/cmd/setCorrelationLogging.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+		cred, err := GetCredentials(setCorrelationLoggingEnvironment)
+		if err != nil {
+			utils.HandleErrorAndExit("Error getting credentials", err)
+		}
+		executeSetCorrelationLoggingCmd(cred)
+	},
+}
+
+func executeSetCorrelationLoggingCmd(credential credentials.Credential) {
+	resp, err := impl.SetCorrelationLoggingComponent(credential, setCorrelationLoggingEnvironment, setCorrelationLoggingComponentName, setCorrelationLoggingEnabled, setCorrelationLoggingDeniedThreads)
```
Comment (ashera96) [inline import-export-cli/cmd/setCorrelationLogging.go:61]: Shall we span this across multiple lines? this line has too many characters IMO

File: import-export-cli/cmd/setCorrelationLogging.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+	if err != nil {
+		utils.Logln(utils.LogPrefixError+"Setting the correlation log component configs", err)
+		utils.HandleErrorAndExit("Error while setting the correlation log component configs", err)
+	}
+	// Print info on response
+	utils.Logf(utils.LogPrefixInfo+"ResponseStatus: %v\n", resp)
```
Comment (ashera96) [inline import-export-cli/cmd/setCorrelationLogging.go:67]:
```suggestion
	utils.Logf(utils.LogPrefixInfo + "ResponseStatus: %v\n", resp)
```

**Review by npamudika — APPROVED** [review]

File: import-export-cli/impl/logger.go [hunk 9]
```diff
@@ -23,6 +23,7 @@ import (
 	"errors"
 	"fmt"
 	"io"
+	"strings"
```
Comment (ashera96) [inline import-export-cli/impl/logger.go:26]: Shouldn't this import come below "os" import?

File: import-export-cli/cmd/getCorrelationLogging.go (hunk 1 above)
Comment (wasuradananjith) [reply import-export-cli/cmd/getCorrelationLogging.go:59]: By default go linter does that automatically. When there is a comma-separated value after a string inside the parenthesis, Go removes the spaces in between other variables. Here [1] also the same thing has happened. Changing this manually will result in changing this back to this again when some other person edits this file.

[1] https://github.com/wso2/product-apim-tooling/blob/master/import-export-cli/cmd/getAPILogging.go#L61

File: import-export-cli/cmd/setCorrelationLogging.go (hunk 2 above)
Comment (wasuradananjith) [reply import-export-cli/cmd/setCorrelationLogging.go:63]: By default go linter does that automatically. When there is a comma-separated value after a string inside the parenthesis, Go removes the spaces in between other variables. Here [1] also the same thing has happened. Changing this manually will result in changing this back to this again when some other person edits this file.

[1] https://github.com/wso2/product-apim-tooling/blob/master/import-export-cli/cmd/getAPILogging.go#L61

File: import-export-cli/impl/logger.go (hunk 3 above)
Comment (wasuradananjith) [reply import-export-cli/impl/logger.go]: Fixed via https://github.com/wso2/product-apim-tooling/pull/966/commits/5584c0cda4fcd40ca92ceb0a8368f41672eba295

File: import-export-cli/impl/logger.go (hunk 4 above)
Comment (wasuradananjith) [reply import-export-cli/impl/logger.go]: Fixed via https://github.com/wso2/product-apim-tooling/pull/966/commits/5584c0cda4fcd40ca92ceb0a8368f41672eba295

File: import-export-cli/impl/logger.go (hunk 5 above)
Comment (wasuradananjith) [reply import-export-cli/impl/logger.go:353]: By default go linter does that automatically. When there is a comma-separated value after a string inside the parenthesis, Go removes the spaces in between other variables. Here [1] also the same thing has happened. Changing this manually will result in changing this back to this again when some other person edits this file.

[1] https://github.com/wso2/product-apim-tooling/blob/master/import-export-cli/cmd/getAPILogging.go#L61

File: import-export-cli/impl/logger.go (hunk 6 above)
Comment (wasuradananjith) [reply import-export-cli/impl/logger.go]: Fixed via https://github.com/wso2/product-apim-tooling/commit/5584c0cda4fcd40ca92ceb0a8368f41672eba295

File: import-export-cli/cmd/setCorrelationLogging.go (hunk 8 above)
Comment (wasuradananjith) [reply import-export-cli/cmd/setCorrelationLogging.go:67]: By default go linter does that automatically. When there is a comma-separated value after a string inside the parenthesis, Go removes the spaces in between other variables. Here [1] also the same thing has happened. Changing this manually will result in changing this back to this again when some other person edits this file.

[1] https://github.com/wso2/product-apim-tooling/blob/master/import-export-cli/cmd/getAPILogging.go#L61

File: import-export-cli/cmd/setCorrelationLogging.go (hunk 7 above)
Comment (wasuradananjith) [reply import-export-cli/cmd/setCorrelationLogging.go]: Fixed via https://github.com/wso2/product-apim-tooling/pull/966/commits/760b88238912fcd2b560f35b8e9ec5ee1f147ce7

File: import-export-cli/impl/logger.go (hunk 9 above)
Comment (wasuradananjith) [reply import-export-cli/impl/logger.go]: Fixed via https://github.com/wso2/product-apim-tooling/pull/966/commits/5584c0cda4fcd40ca92ceb0a8368f41672eba295

**Review by ashera96 — APPROVED** [review]

**Pull request merged**